### ISSUE TITLE
Support PHP7 type hints for action params.

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/yii2#10836]
With PHP7 comes type hinting for basic types like int and string.
Yii should support these type hints for action parameters. 
For example:
- Convert numeric strings to appropriate type based on type hint.
- Throw exception on invalid data just like it does now for array type hints.

This issue is just to have it written down. I am not yet on PHP7 so don't intend to implement this myself.

**Comment by cebe** [comment]
dependend on strict mode PHP might already do this. Anyway it should be verified at least.

**Comment by SamMousa** [comment]
As far as I can tell strict mode will throw an error for any non-string type hint since all query parameters are string. So then we would get errors when actually passing numbers: `?id=12345` would throw errors while ideally we want to get an `int`.

**Comment by Faryshta** [comment]
wouldn't this break compatibility with any previous version of php?

**Comment by SamMousa** [comment]
@Faryshta No it would not. PHP7 just has some extra typehints that can be used to help with parameter binding. This is transparent to the user and can be done in such a way it works for PHP7 without influencing PHP <7

**Comment by Faryshta** [comment]
@SamMousa can you provide a code example in 3v4l.org?

**Comment by SamMousa** [comment]
https://3v4l.org/Yfd0f

Basically in PHP < 7, any code that uses scalar type hints won't work. However we can still use the same reflection to check code that works and apply conversions based on that.

So if as a developer I use PHP7, I could use:
````
public function actionRead(int $id) {
...
}
````
And if I target PHP < 7, I would use (as I do today):
````
public function actionRead($id) {
...
}
````

Both scenarios can be supported (read: injected / converted) without PHP7 specific code.

**Comment by SamMousa** [comment]
@cebe, you are right about that still we would want to do the conversion even when strict mode is enabled?
If not then you can never use type hints and strict mode together since query params are always strings.
http://php.net/manual/en/migration70.new-features.php

**Comment by Faryshta** [comment]
i don't think that code is feasible. I would prefer to ask php7 as the minium version for 2.1

**Comment by SamMousa** [comment]
I don't see how that's related; what code is not feasible exactly?

You are referring to the conditional class definition.. That is just for
illustration and would not be part of yii.

**Comment by cebe** [comment]
> still we would want to do the conversion even when strict mode is enabled?

yes sure, that would be the expected behaviour.

**Comment by SamMousa** [comment]
Looking into this some more now as I've made the switch to PHP7.

Currently Yii code will continue to work regardless of `strict_types` for several reasons.
1. `strict_types` applies to function calls not declarations in a file, so as long as Yii doesn't include it in the core it;ll be fine.
2. There is a PHP bug that causes the `strict_types` directive not to apply to `call_user_func` and family when used in side a namespace: https://bugs.php.net/bug.php?id=72248

Regarding using `strict_types` in the core: PHP < 7 will throw a warning that I have been unable to hide. So I agree with @Faryshta that at least the core should not include `strict_types` until the php requirement is `<= 7`.

For now controller actions that use php 7 type hints will work in coercive mode.

Feel free to close this or leave it open for reference.

**Comment by dynasource** [comment]
> Feel free to close this

**Comment by Faryshta** [comment]
can we discuss this again focusing on the 2.1 branch?
